### PR TITLE
Add better exception granurality to memory::alloc_shared and memory::alloc_unique

### PR DESCRIPTION
**Pull request by JanuszL**
- makes memory::alloc_shared and memory::alloc_unique to use CUDABadAlloc for GPU and  pinned memory allocation instead of generic std::bad_alloc

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It adds better exception granurality to memory::alloc_shared and memory::alloc_unique

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     makes memory::alloc_shared and memory::alloc_unique to use CUDABadAlloc for GPU and  pinned memory allocation instead of generic std::bad_alloc
 - Affected modules and functionalities:
     alloc.cc
 - Key points relevant for the review:
     NA
 - Validation and testing:
     tests are added
 - Documentation (including examples):
     NA

**JIRA TASK**: *[NA]*

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

**Review by mzient — APPROVED** [review]

**Comment by JanuszL** [comment]
!build

File: dali/kernels/alloc.cc [hunk 1]
```diff
@@ -23,6 +23,16 @@ namespace dali {
...
 namespace memory {
 
+void ThrowMemoryError(AllocType type, size_t requested_size) {
+  if (type == AllocType::Pinned) {
+    throw dali::CUDABadAlloc(requested_size, true);
+  } else if (type == AllocType::GPU) {
+    throw dali::CUDABadAlloc(requested_size, false);
+  } else {
+    throw std::bad_alloc();
+  }
```
Comment (mzient) [inline dali/kernels/alloc.cc:33]:
```suggestion
  if (type != AllocType::Host) {
    auto err = cudaGetLastError();
    if (err == cudaErrorMemoryAllocation)
       throw dali::CUDABadAlloc(requested_size, type == AllocType::Pinned);
    else
       throw dali::CUDAError(err);
  } else {
    throw std::bad_alloc();
  }
```

**Comment by JanuszL** [comment]
!build

File: include/dali/core/dev_buffer.h [hunk 2]
```diff
@@ -156,7 +156,7 @@ struct DeviceBuffer {
     T *ptr = nullptr;
     CUDA_CALL(cudaMalloc(reinterpret_cast<void**>(&ptr), count * sizeof(T)));
     if (!ptr)
-      throw std::bad_alloc();
+      throw dali::CUDABadAlloc(count * sizeof(T));
```
Comment (mzient) [inline include/dali/core/dev_buffer.h:159]:
```suggestion
    if (!ptr) {
      (void)cudaGetLastError();
      throw dali::CUDABadAlloc(count * sizeof(T));
    }
```

File: dali/kernels/alloc.cc (hunk 1 above)
Comment (JanuszL) [reply dali/kernels/alloc.cc]: Done

File: include/dali/core/dev_buffer.h (hunk 2 above)
Comment (JanuszL) [reply include/dali/core/dev_buffer.h]: Done

**Comment by JanuszL** [comment]
!build

**Review by mzient — APPROVED** [review]

**Review by banasraf — APPROVED** [review]

**Pull request merged**